modal/shared/logging.py

```diff
@@ -0,0 +1,107 @@
+
+
+def get_logger(name: str):
+    return logging.getLogger(name)
+
+
+# Define a custom logging handler that sends logs to Datadog
+class DatadogHandler(logging.Handler):
+    def __init__(self, api_client: ApiClient):
+        self.api_client = api_client
+
+        super().__init__()
+
+    def emit(self, record):
+        # Ignore debug messages
+        if record.levelno == logging.DEBUG:
+            return
+
+        environment_name = os.getenv("DD_ENV") or "development"
+
+        log_payload = {
+            "python-logging": {
+                "py-env": "development",
+                "py-message": record.getMessage(),
+                "py-status": record.levelname.lower(),
+                "py-logger": record.name,
+                "py-stacktrace": record.exc_info,
+                "py-exception": record.exc_text,
+                "py-line": record.lineno,
+                "py-file": record.filename,
+                "py-function": record.funcName,
+                "py-level": record.levelno,
+                "py-path": record.pathname,
+                "py-thread": record.thread,
+                "py-threadName": record.threadName,
+                "py-process": record.process,
+                "py-processName": record.processName,
+                "py-args": record.args,
+                "py-msecs": record.msecs,
+                "py-relativeCreated": record.relativeCreated,
+                "py-created": record.created,
+                "py-module": record.module,
+            },
+            "message": record.getMessage(),
+            "environment": environment_name,
+            "source": record.name,
+            "model": record.__dict__.get("model"),
+            "modal": {
+                "cloudProvider": os.getenv("MODAL_CLOUD_PROVIDER") or "",
+                "environment": os.getenv("MODAL_ENVIRONMENT") or "",
+                "imageId": os.getenv("MODAL_IMAGE_ID") or "",
+                "region": os.getenv("MODAL_REGION") or "",
+                "taskId": os.getenv("MODAL_TASK_ID") or "",
+            },
+        }
+
+        # Send the log to Datadog using the Logs API
+        try:
+            body = HTTPLog(
+                [
+                    HTTPLogItem(
+                        ddsource="Python",
+                        ddtags=f"env:{environment_name}",
+                        hostname=os.getenv("MODAL_TASK_ID") or "",
+                        message=json.dumps(log_payload),
+                        service="openrouter-runner",
+                        status=record.levelname.lower(),
+                    ),
+                ]
+            )
+
+            logs = LogsApi(api_client)
+
+            logs.submit_log(
+                content_encoding=ContentEncoding.DEFLATE,
+                body=body,  # type: ignore
+            )
+
+        except Exception as e:
+            print(f"Error sending log to Datadog: {e}")
+
+
+if "DD_API_KEY" in os.environ:
```

Comment:
given the new docs say:
> or leave it blank to disable Datadog (e.g. `DD_API_KEY=`

do you know if `DD_API_KEY` will still be on the container as empty string if ppl set it like that?

if so, would probs need to be
```suggestion
if os.environ.get("DD_API_KEY") is not None:
```

Reply:
Doing os-environ check at the top-level is a bit risky here as well, because the top-level scope might not have the env injected I think?

Maybe we should make this into an `init_logger` function and call it at the top of each stub :-?

Reply:
mm yeah thats a great point.. will look kinda odd but feels like a reasonable convention to stick to given the modal quirkyness

Reply:
Great catch on DD_API_KEY as an empty string, fixed it!

For the env var access timing, I've tested it a bunch as is and it works as expected. Is there a certain situation you think it might break in? The environment environment variables should (and appear to be) present whenever the code is interpreted. If we nest this inside a function call, there will no longer be a certain that a logger returned by `get_logger` will be configured to send immediately